### ISSUE TITLE
interface Package

### ISSUE THREAD

**Issue by gertcuykens** [microsoft/language-server-protocol#878]
Trying to figure out how a outline for programing languages that are package based, like `Go` can make a package outline request that exist out of several documents source files? Example of a package outline https://golang.org/pkg/encoding/json/

Looking for something like this I guess?
```
interface Package {
	src: []Document;
}

interface Document {
	uri: DocumentUri;
	symbols: []DocumentSymbol;
}
```

**Comment by dbaeumer** [comment]
Resolving the documents per package needs to be a custom request since LSP has no support for this.

The outline can be queried using the document symbol request.

Note that LSP clearly states that request are valid even if the document is not open (no open notification has been sent).

**Comment by gertcuykens** [comment]
But then @stamblerre stated that needed on the gopls side is the ability to send back a URI for DocumentSymbol to do this https://github.com/microsoft/vscode-go/issues/2793#issuecomment-566424239 but that isn't allowed in lsp either? microsoft/language-server-protocol#582

**Comment by dbaeumer** [comment]
I am still not sure why you would need to send back the URI. The URI IMO would always be the one that is used in the request params. If not could you provide me with an example where that differs?

**Comment by gertcuykens** [comment]
@stamblerre @ramya-rao-a Can you point to the http outline request in vscode-go please where URI is needed to be able to build a complete outline package tree and not just the outline of a single document, Thanks

**Comment by stamblerre** [comment]
The URI would be needed if `DocumentSymbol` were to be the response for some new request (e.g. `textDocument/packageSymbols`, because then you would need to be able to send back `DocumentSymbols` for every URI in the package. It doesn't seem like the LSP spec will support this, and we don't have any plan to support a custom request like this in `gopls`. As it is, I think it would be fairly simple for a client to request `textDocument/documentSymbol` for every file in a directory as a rough approximation of such a request.

@gertcuykens: If this is still an issue that you feel strongly about, I think the best course of action would be to open a new issue on the [Go issue tracker](https://github.com/golang/go/issues/new) and to give an example of how it would work and how the UI would look.

Thanks for the responses, @dbaeumer.

**Comment by gertcuykens** [comment]
Thanks for response. PS That was the first thing I did but it was closed as duplicated https://github.com/microsoft/vscode-go/issues/2910 which referred to https://github.com/microsoft/vscode-go/issues/2793 as an open discussion and was blocked because of this issue we are now in. I understand if your opinion is that godoc is better for a package outline tree but I was under the impression you were in favor but blocked by the lsp spec? It's ok, I just want to make sure if the spec is not the issue anymore for a potential implementation for this feature. I am not familiar with the code base so I rely on your technical feedback before me or somebody else would make an attempt to implement it without shooting myself in the foot because of something I fail to understand.

**Comment by stamblerre** [comment]
Sorry for not being clear earlier - I thought maybe there could be a possibility to modify the LSP spec to support a feature like this, but it sounds like that won't be an option. If the only alternative is a custom request - that's not something we're ready to support in `gopls` yet.

**Comment by dbaeumer** [comment]
I am not in principal against there requests, but we need to keep the following things in mind:

- is this something a language server or a project server answers
- how can different languages profit from this 
- IMO it must allow for lazy evaluation. First get the package structure and then the outline for an expanded element in the package structure.

**Comment by Gama11** [comment]
>is this something a language server or a project server answers

What's a "project server"? Is there some "Project Server Protocol" I'm not aware of? :)

**Comment by dbaeumer** [comment]
Not officially but there were discussion what project system, build system (there is a BSP) and a language server should do.

**Comment by gertcuykens** [comment]
Maybe I am comparing apples with oranges here but I assume not a single language consist of just one file and import other files in one way or another that are scoped under one name? Granted there are many languages that are used for simple projects where it's sufficient that you can click on every file to have a outline of that particular file. But almost all language have a sort of ctags tool to link different documents together for a more global structure, I would even say ctags standard would be considered one of the first language protocols. Is it the task of the language server or the client to create this structure? Perhaps not the server, but definitely should be made into a standard one way the other.
